influxql/engine.go

```diff
@@ -1,1250 +1,554 @@
+			if vals[i] != nil {
+				return false
+			}
+		}
 	}
+	return true
 }
 
-type meanMapOutput struct {
-	Count int
-	Sum   float64
-}
-
-// ReduceMean computes the mean of values for each key.
-func ReduceMean(key Key, values []interface{}, e *Emitter) {
-	out := &meanMapOutput{}
-	for _, v := range values {
-		val := v.(*meanMapOutput)
-		out.Count += val.Count
-		out.Sum += val.Sum
-	}
-	if out.Count > 0 {
-		e.Emit(key, out.Sum/float64(out.Count))
-	}
-}
+// processRawResults will handle converting the reduce results from a raw query into a Row
+func (m *MapReduceJob) processRawResults(resultValues [][]interface{}) *Row {
+	selectNames := m.stmt.NamesInSelect()
 
-// MapMin collects the values to pass to the reducer
-func MapMin(itr Iterator, e *Emitter, tmin int64) {
-	var min float64
-	pointsYielded := false
-
-	for k, _, v := itr.Next(); k != 0; k, _, v = itr.Next() {
-		val := v.(float64)
-		// Initialize min
-		if !pointsYielded {
-			min = val
-			pointsYielded = true
+	// ensure that time is in the select names and in the first position
+	hasTime := false
+	for i, n := range selectNames {
+		if n == "time" {
+			if i != 0 {
+				tmp := selectNames[0]
+				selectNames[0] = "time"
+				selectNames[i] = tmp
+			}
+			hasTime = true
 		}
-		min = math.Min(min, val)
-	}
-	if pointsYielded {
-		e.Emit(Key{tmin, itr.Tags()}, min)
 	}
-}
 
-// ReduceMin computes the min of value.
-func ReduceMin(key Key, values []interface{}, e *Emitter) {
-	var min float64
-	pointsYielded := false
-
-	for _, v := range values {
-		val := v.(float64)
-		// Initialize min
-		if !pointsYielded {
-			min = val
-			pointsYielded = true
-		}
-		m := math.Min(min, val)
-		min = m
-	}
-	if pointsYielded {
-		e.Emit(key, min)
+	// time should always be in the list of names they get back
+	if !hasTime {
+		selectNames = append([]string{"time"}, selectNames...)
 	}
-}
 
-// MapMax collects the values to pass to the reducer
-func MapMax(itr Iterator, e *Emitter, tmax int64) {
-	var max float64
-	pointsYielded := false
-
-	for k, _, v := itr.Next(); k != 0; k, _, v = itr.Next() {
-		val := v.(float64)
-		// Initialize max
-		if !pointsYielded {
-			max = val
-			pointsYielded = true
-		}
-		max = math.Max(max, val)
-	}
-	if pointsYielded {
-		e.Emit(Key{tmax, itr.Tags()}, max)
-	}
-}
+	// if they've selected only a single value we have to handle things a little differently
+	singleValue := len(selectNames) == 2
 
-// ReduceMax computes the max of value.
-func ReduceMax(key Key, values []interface{}, e *Emitter) {
-	var max float64
-	pointsYielded := false
-
-	for _, v := range values {
-		val := v.(float64)
-		// Initialize max
-		if !pointsYielded {
-			max = val
-			pointsYielded = true
-		}
-		max = math.Max(max, val)
+	row := &Row{
+		Name:    m.MeasurementName,
+		Tags:    m.TagSet.Tags,
+		Columns: selectNames,
 	}
-	if pointsYielded {
-		e.Emit(key, max)
+
+	// return an empty row if there are no results
+	if len(resultValues) == 0 || len(resultValues[0]) != 2 {
```

Comment:
This magic number isn't that helpful. Is there a better way to communicate what is going on here?